### PR TITLE
Latest Posts: add screen reader title text to Read more links and use alternative to excerpt_more filter

### LINKED ISSUE #55026: Latest Posts: Read more links have identical text and ignore `excerpt_more` filter from themes
### Description

When #51190 matched the front end to the editor, instead of matching the editor to the front end, it created two problems:
1. All "Read more" links in the block have identical text, which goes [against accessibility guidelines](https://make.wordpress.org/themes/handbook/review/accessibility/required/#repetitive-link-text).
2. By using the `excerpt_more` hook, on the standard priority, this replaces filters that themes have used to add custom unique links in place of the ellipsis. That includes the bundled themes from Twenty Thirteen to Twenty Nineteen and Twenty Twenty-One.

Related: [Trac 59409](https://core.trac.wordpress.org/ticket/59409)

### Step-by-step reproduction instructions

1. Create several posts, each with a few paragraphs of text.
2. Add a Latest Posts block, either in a template or within the content of a page. In the sidebar, toggle the Post content on and show the Excerpt.
3. Activate Twenty Twenty-Three and visit a page with that Latest Posts block.
4. Observe that the post excerpts end with the ellipsis (…) and "Read more" links.
5. Check the text of the "Read more" links, and note that their accessible names are identical to each other. (You could inspect each element, or you might access the list of links with a screen reader or in a tool such as the WAVE browser extension.)
6. Activate Twenty Twenty-One and return to the page you visited earlier.
7. Check the text of "Read more" links again. With WordPress 6.2, these links showed "Continue reading" and included the post title as screen-reader-only text. In WordPress 6.3, that was replaced with the ellipsis and "Read more" links.

### Screenshots, screen recording, code snippet

Twenty Twenty-Three had the bracketed ellipsis in WordPress 6.2

ellipsis without a link

WordPress 6.3 added "Read more" links to each post excerpt

'Read more' links with Twenty Twenty-Three

Twenty Twenty-One's "Continue reading" links applied to the post excerpts with the Latest Posts block in WordPress 6.2

'Continue reading' links

WordPress 6.3 replaced the theme's text with "Read more" links

'Read more' links with Twenty Twenty-One

### Environment info

WordPress 6.3, with or without Gutenberg 16.7.0

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by ramonjd**
Maybe resolves https://github.com/WordPress/gutenberg/issues/55026

Follow up to https://github.com/WordPress/gutenberg/pull/51190

## What?

**In the editor**: adds a screen reader text span with the post title in the i18n interpolator

**In the frontend**: removes `excerpt_more` filter so we don't override themes and also replaces the default ellipsis with an accessible read more link. This is done only when the Latest Post attribute `displayPostContentRadio` is explicitly set to `excerpt`.

Also I notices that the frontend didn't match the editor in one respect: the "Read more" default text should only appear when the attribute `excerptLength` is less than the currently-set excerpt length (from any filters etc)

❗ It's important to note that themes, such as 2019, that specific a custom excerpt more text will not have that custom text and link reflected in the editor when the `displayPostContentRadio` is set to `excerpt` and the attribute `excerptLength` is less than the currently-set excerpt length. The editor's preferences take precedence here. . "Read more" will always show in this case.

## Why?
Use of the `excerpt_more` filter was overriding theme defaults, e.g., 2021, 2019

The repetition of "Read more" text in the frontend and editor did not provide any context to screen reader users. SeE: https://make.wordpress.org/themes/handbook/review/accessibility/required/#repetitive-link-text

## Testing Instructions

1. Create a few posts with text content.
2. Create another new post in a block theme and add a Latest Posts block
3. In the block setting sidebar, activate Post content > Excerpts radio button. Slide the "MAX NUMBER OF WORDS" so that the "Read more" link appears.
4. Save the post and preview in the frontend
5. Ensure the "Read more" link is consistent between editor and frontend - check that the screen reader text with the title appears in the source HTML
6. Check in other themes, especially 2021, 2019
7. Ensure that theme excerpt more copy takes precedence on the frontend. (2021 might show an extra ellipsis in the editor - it's inconsistent between themes, so I was aiming for the most general approach)
8. Play around with the "MAX NUMBER OF WORDS" attribute in various theme.

#### Editor

Screenshot 2023-10-04 at 5 35 25 pm

#### Frontend
Screenshot 2023-10-04 at 5 35 30 pm

### CODE REVIEW

File: packages/block-library/src/latest-posts/index.php [hunk 1]
```diff
@@ -48,14 +48,6 @@ function render_block_core_latest_posts( $attributes ) {
 	$block_core_latest_posts_excerpt_length = $attributes['excerptLength'];
 	add_filter( 'excerpt_length', 'block_core_latest_posts_get_excerpt_length', 20 );
 
-	$filter_latest_posts_excerpt_more = static function ( $more ) use ( $attributes ) {
-		$use_excerpt = 'excerpt' === $attributes['displayPostContentRadio'];
```
Comment (ramonjd) [inline packages/block-library/src/latest-posts/index.php:52]: As mentioned in https://github.com/WordPress/gutenberg/issues/55026#issuecomment-1746246347

I think I used this so that choices of the user in the editor were reflected on the frontend, even if those changes overrode theme custom text.

Theme custom excerpt text would still display if displayPostContentRadio was not set to 'excerpt' in the Latest Posts block attributes. However, I believe we were missing the editor's "excerptLength attribute is less than the currently-set excerpt length" [condition](https://github.com/WordPress/gutenberg/blob/4b804ac39b34332115562817c717ff4b7206c36e/packages/block-library/src/latest-posts/edit.js#L476-L476) in index.php.

What should prevail here?

File: packages/block-library/src/latest-posts/index.php [hunk 2]
```diff
@@ -151,6 +143,24 @@ function render_block_core_latest_posts( $attributes ) {
 
 			$trimmed_excerpt = get_the_excerpt( $post );
 
+			/*
+			 * Adds a "Read more" link with screen reader text.
+			 * [&hellip;] is the default excerpt ending from wp_trim_excerpt() in Core.
+			 */
+			if ( str_ends_with( $trimmed_excerpt, ' [&hellip;]' ) && 'excerpt' === $attributes['displayPostContentRadio'] ) {
```
Comment (andrewserong) [inline packages/block-library/src/latest-posts/index.php]: I haven't tested this PR yet, so this is just a drive-by comment! 😄

The condition above appears to already check for `&& 'excerpt' === $attributes['displayPostContentRadio'] )` so I don't think we need that condition in this `if` statement?

Comment (ramonjd) [reply packages/block-library/src/latest-posts/index.php]: Did I leave that in there? 😅 

Thanks for spotting 🙇🏻

Comment (ramonjd) [reply packages/block-library/src/latest-posts/index.php]: One more comment here: I'm not 100% convinced of the robustness of this check. We're hard-coding what we expect the default value to be (from Core). 

If a theme has added a hook that returns ` [&hellip;]`, which we presume is what they want, then this condition will also overwrite.

File: packages/block-library/src/latest-posts/index.php [hunk 3]
```diff
@@ -151,6 +143,24 @@ function render_block_core_latest_posts( $attributes ) {
...
 
+			/*
+			 * Adds a "Read more" link with screen reader text.
+			 * [&hellip;] is the default excerpt ending from wp_trim_excerpt() in Core.
+			 */
+			if ( str_ends_with( $trimmed_excerpt, ' [&hellip;]' ) ) {
+				$excerpt_length = (int) apply_filters( 'excerpt_length', $block_core_latest_posts_excerpt_length );
+				if ( $excerpt_length <= $block_core_latest_posts_excerpt_length ) {
+					$trimmed_excerpt  = substr( $trimmed_excerpt, 0, -11 );
+					$trimmed_excerpt .= sprintf(
```
Comment (andrewserong) [inline packages/block-library/src/latest-posts/index.php:154]: This is fairly subtle, but should we run the output of `sprintf()` through `wp_kses_post` since the first param passed to `sprintf` is a translated string, rather than one we're in direct control of within this function?

File: packages/block-library/src/latest-posts/index.php (hunk 1 above)
Comment (andrewserong) [reply packages/block-library/src/latest-posts/index.php:52]: Great question. The way you're currently handling it in this PR feels natural to me in testing!

File: packages/block-library/src/latest-posts/edit.js [hunk 4]
```diff
@@ -483,19 +483,27 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
 								.split( ' ', excerptLength )
 								.join( ' ' ) }
 							{ createInterpolateElement(
-								/* translators: excerpt truncation character, default …  */
-								__( ' … <a>Read more</a>' ),
+								sprintf(
+									/* translators: %1$s is a static string "Read more", and %2$s is the post title only visible to screen readers. */
+									__( '… <a>%1$s<span>: %2$s</span></a>' ),
+									__( 'Read more' ),
+									titleTrimmed
```
Comment (andrewserong) [inline packages/block-library/src/latest-posts/edit.js]: Does this need a fallback like we use for displaying the title? I.e.

```suggestion
									titleTrimmed || __( '(no title)' )
```

**Review by andrewserong — COMMENTED** [review]
This is testing nicely for me @ramonjd and feels like a natural way to handle it while preserving themes that intentionally override `excerpt_more` 👍

✅ Themes that override the excerpt now output correctly on the site frontend (tested 2019, and 2021)
✅ Themes that don't override the excerpt correctly output the `Read more` text along with the screen reader text containing the post title
❓ If there is no post title, then that's flagged in the screen reader text, too on the site frontend. E.g. the following:

image

However in the editor, it looks like the fallback `(no title)` isn't being output in the screen reader text. Just left a small comment there and one for the output of the PHP, but otherwise this is looking good to me!

| 2019 before | 2019 after |
| --- | --- |
| image | image |

File: packages/block-library/src/latest-posts/edit.js [hunk 5]
```diff
@@ -483,19 +483,27 @@ export default function LatestPostsEdit( { attributes, setAttributes } ) {
 								.split( ' ', excerptLength )
 								.join( ' ' ) }
 							{ createInterpolateElement(
-								/* translators: excerpt truncation character, default …  */
-								__( ' … <a>Read more</a>' ),
+								sprintf(
+									/* translators: %1$s is a static string "Read more", and %2$s is the post title only visible to screen readers. */
```
Comment (costdev) [inline packages/block-library/src/latest-posts/edit.js]:
```suggestion
									/* translators: 1: The static string "Read more", 2: The post title only visible to screen readers. */
```
[See here](https://github.com/WordPress/wordpress-develop/blob/6.3/src/wp-admin/includes/class-theme-upgrader.php#L83) for examples of how to document multiple `sprintf()` tokens.

File: packages/block-library/src/latest-posts/index.php (hunk 3 above)
Comment (costdev) [reply packages/block-library/src/latest-posts/index.php:154]: @andrewserong Core generally trusts its translations so we tend not to run them through escaping functions. Since the translated string uses the default domain rather than a `'gutenberg'` domain, it should likely be treated the same as any other Core translation.

Comment (andrewserong) [reply packages/block-library/src/latest-posts/index.php:154]: Excellent, thanks for confirming @costdev! 🙂

I'll click "resolve" on this thread.

File: packages/block-library/src/latest-posts/index.php [hunk 6]
```diff
@@ -151,6 +143,24 @@ function render_block_core_latest_posts( $attributes ) {
 
 			$trimmed_excerpt = get_the_excerpt( $post );
 
+			/*
+			 * Adds a "Read more" link with screen reader text.
+			 * [&hellip;] is the default excerpt ending from wp_trim_excerpt() in Core.
+			 */
+			if ( str_ends_with( $trimmed_excerpt, ' [&hellip;]' ) ) {
```
Comment (costdev) [inline packages/block-library/src/latest-posts/index.php:150]: @ramonjd Would this be better?

```php
/** This filter is documented in wp-includes/formatting.php */
if ( str_ends_with( $trimmed_excerpt, apply_filters( 'excerpt_more', ' ' . '[&hellip;]' ) ) {
```
[Reference](https://developer.wordpress.org/reference/hooks/excerpt_more/)

**Comment by ramonjd** [comment]
Thanks @andrewserong and @costdev for helping make this PR better. I think I've addressed all feedback.

**Review by andrewserong — APPROVED** [review]
Thanks for the updates, just re-tested and this is all looking good to me now! ✨

**Comment by mikachan** [comment]
I just cherry-picked this PR to the 6.4-beta3 branch to get it included in the next release: https://github.com/WordPress/gutenberg/pull/55039/commits/11f6a6d7a885994c11dc5863239f19b1e923642e
